primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +651,249 @@ where
+            let stream = warmup_new_stream().unwrap();
+
+            #[cfg(feature = "kzg-print-trace")]
+            let commit_time =
+                start_timer!(|| format!("Committing to polynomial of degree {} ", poly.degree()));
+
+            let mut srs_on_gpu = Self::load_prover_param_to_gpu(prover_param, poly.degree())?;
+            let poly_on_gpu = Self::load_poly_to_gpu(poly)?;
+            let msm_result_on_gpu = Self::commit_on_gpu(&mut srs_on_gpu, &poly_on_gpu, &stream)?;
+            let comm = Self::load_commitment_to_host(msm_result_on_gpu, &stream)?;
+
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(commit_time);
+
+            Ok(comm)
+        }
+
+        /// type conversion (specializable/overridable for concrete types) from
+        /// field in arkworks to icicle
+        ///
+        /// # NOTE
+        /// returned icicle field is in normal affine (not montgomery) form
+        fn ark_field_to_icicle(f: E::ScalarField) -> C::ScalarField {
+            // this is more memory efficient than default `.from_ark`
+            C::ScalarField::from_bytes_le(&f.into_bigint().to_bytes_le())
+        }
+
+        /// type conversion (specializable/overridable for concrete types) from
+        /// affine point in arkworks to icicle
+        ///
+        /// # NOTE
+        /// returned icicle field is in normal affine (not montgomery) form
+        fn ark_affine_to_icicle(p: E::G1Affine) -> IcicleAffine<C> {
+            // this is more memory efficient than default `.from_ark`
+            IcicleAffine {
+                x: C::BaseField::from_bytes_le(&p.x.into_bigint().to_bytes_le()),
+                y: C::BaseField::from_bytes_le(&p.y.into_bigint().to_bytes_le()),
+            }
+        }
+
+        /// load SRS for prover onto GPU once, and reuse for future poly-commit
+        /// of degree<=`supported_degree`
+        fn load_prover_param_to_gpu<'srs>(
+            prover_param: impl Borrow<UnivariateProverParam<E>>,
+            supported_degree: usize,
+        ) -> Result<HostOrDeviceSlice<'srs, IcicleAffine<C>>, PCSError> {
+            let prover_param = prover_param.borrow();
+            if supported_degree > prover_param.powers_of_g.len() - 1 {
+                return Err(PCSError::InvalidParameters(format!(
+                    "supported degree {} is larger than allowed {}",
+                    supported_degree,
+                    prover_param.powers_of_g.len() - 1
+                )));
+            }
+
+            let mut bases_on_device =
+                HostOrDeviceSlice::<'_, IcicleAffine<C>>::cuda_malloc(supported_degree + 1)?;
+
+            #[cfg(feature = "kzg-print-trace")]
+            let conv_time = start_timer!(|| "Type Conversion: ark->ICICLE: Group");
+            let bases: Vec<IcicleAffine<C>> = parallelizable_slice_iter(&prover_param.powers_of_g[..supported_degree + 1])
+                    .map(|&p| Self::ark_affine_to_icicle(p)) // TODO: maybe add api?
```

Comment:
Yes, is that a problem? It's so fast now, I don't see it as a concern. And we don't have unsafe or implicit "forced type interpretation" happening here.

Reply:
It could be a (small) problem when we are committing large amounts of data.
E.g, for 32MB payload data would be encoded into 2^{17~18} field elements.

Reply:
Also if we are to use it for plonk proof generation. Often you need to deal with an even larger degree.

Reply:
BTW it's not urgent, you can leave it as an issue

Reply:
so you would suggest: 
load the arkwork type byte-by-byte to GPU, then let GPU directly interpret that on the GPU?

my 2 cents is the `msm()` rust bindings will have rust's type system checks anyway (meaning `msm` is expecting `IcicleAffine<IcicleCurve>`, we might as well just raw pointer cast it in CPU. 
(unless we just completely go around ICICLE's `msm` api and write our own logic using their `msm` via FFI.